File: castplay-allinone/android/app/src/main/java/com/castplay/player/ErrorReporter.kt
package com.castplay.player

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * 错误上报服务
 * 统一收集和上报 Android 播放端错误
 */
class ErrorReporter private constructor(private val context: Context) {

    companion object {
        private const val TAG = "ErrorReporter"

        // 错误类型常量
        const val TYPE_DOWNLOAD_FAILED = "download_failed"
        const val TYPE_PLAYBACK_ERROR = "playback_error"
        const val TYPE_CACHE_ERROR = "cache_error"
        const val TYPE_NETWORK_ERROR = "network_error"
        const val TYPE_PLAYLIST_SYNC_FAILED = "playlist_sync_failed"
        const val TYPE_REGISTRATION_FAILED = "registration_failed"
        const val TYPE_INSUFFICIENT_STORAGE = "insufficient_storage"
        const val TYPE_SWITCH_FAILED = "switch_failed"
        const val TYPE_MEDIA_LOAD_ERROR = "media_load_error"
        const val TYPE_UNKNOWN_ERROR = "unknown_error"

        // 配置
        private const val MAX_QUEUE_SIZE = 50
        private const val FLUSH_INTERVAL_MS = 30000L  // 30 秒
        private const val API_ENDPOINT = "/api/player/devices/notifications"

        @Volatile
        private var instance: ErrorReporter? = null

        fun getInstance(context: Context): ErrorReporter {
            return instance ?: synchronized(this) {
                instance ?: ErrorReporter(context.applicationContext).also { instance = it }
            }
        }
    }

    private val errorQueue = ConcurrentLinkedQueue<ErrorReport>()
    private val executor = Executors.newSingleThreadScheduledExecutor()
    private val prefs by lazy {
        context.getSharedPreferences("error_reporter", Context.MODE_PRIVATE)
    }

    private var deviceId: String? = null
    private var serverUrl: String? = null

    data class ErrorReport(
        val type: String,
        val message: String,
        val playlistId: Long? = null,
        val mediaId: String? = null,
        val timestamp: Long = System.currentTimeMillis(),
        val additionalData: Map<String, Any>? = null
    )

    init {
        // 启动定时刷新任务
        startFlushTimer()
    }

    /**
     * 初始化错误上报服务
     */
    fun init(deviceId: String, serverUrl: String) {
        this.deviceId = deviceId
        this.serverUrl = serverUrl
        Log.d(TAG, "ErrorReporter initialized for device: $deviceId")
    }

    /**
     * 设置设备 ID
     */
    fun setDeviceId(deviceId: String) {
        this.deviceId = deviceId
    }

    /**
     * 设置服务器 URL
     */
    fun setServerUrl(url: String) {
        this.serverUrl = url
    }

    /**
     * 上报错误
     */
    fun reportError(
        type: String,
        message: String,
        playlistId: Long? = null,
        mediaId: String? = null,
        additionalData: Map<String, Any>? = null
    ) {
        if (deviceId == null) {
            Log.w(TAG, "Cannot report error: device ID not set")
            return
        }

        val report = ErrorReport(
            type = type,
            message = message,
            playlistId = playlistId,
            mediaId = mediaId,
            additionalData = additionalData
        )

        // 防止队列过大
        if (errorQueue.size >= MAX_QUEUE_SIZE) {
            errorQueue.poll()  // 移除最旧的
        }

        errorQueue.offer(report)
        Log.w(TAG, "Error reported: [$type] $message")

        // 如果是严重错误，立即发送
        if (isCriticalError(type)) {
            flush()
        }
    }

    /**
     * 判断是否为严重错误
     */
    private fun isCriticalError(type: String): Boolean {
        return type in listOf(
            TYPE_INSUFFICIENT_STORAGE,
            TYPE_REGISTRATION_FAILED,
            TYPE_SWITCH_FAILED
        )
    }

    /**
     * 上报下载失败
     */
    fun reportDownloadFailed(
        message: String,
        playlistId: Long? = null,
        mediaId: String? = null
    ) {
        reportError(TYPE_DOWNLOAD_FAILED, message, playlistId, mediaId)
    }

    /**
     * 上报播放错误
     */
    fun reportPlaybackError(
        message: String,
        mediaId: String? = null,
        additionalData: Map<String, Any>? = null
    ) {
        reportError(TYPE_PLAYBACK_ERROR, message, mediaId = mediaId, additionalData = additionalData)
    }

    /**
     * 上报缓存错误
     */
    fun reportCacheError(message: String) {
        reportError(TYPE_CACHE_ERROR, message)
    }

    /**
     * 上报网络错误
     */
    fun reportNetworkError(message: String) {
        reportError(TYPE_NETWORK_ERROR, message)
    }

    /**
     * 上报播放列表同步失败
     */
    fun reportPlaylistSyncFailed(message: String, playlistId: Long? = null) {
        reportError(TYPE_PLAYLIST_SYNC_FAILED, message, playlistId = playlistId)
    }

    /**
     * 上报存储空间不足
     */
    fun reportInsufficientStorage(requiredBytes: Long, availableBytes: Long) {
        reportError(
            TYPE_INSUFFICIENT_STORAGE,
            "Storage insufficient: required $requiredBytes, available $availableBytes",
            additionalData = mapOf(
                "required_bytes" to requiredBytes,
                "available_bytes" to availableBytes
            )
        )
    }

    /**
     * 刷新错误队列到服务器
     */
    fun flush() {
        if (errorQueue.isEmpty() || deviceId == null || serverUrl == null) {
            return
        }

        val errorsToSend = mutableListOf<ErrorReport>()
        while (errorQueue.isNotEmpty()) {
            errorQueue.poll()?.let { errorsToSend.add(it) }
        }

        if (errorsToSend.isEmpty()) return

        // 在后台线程发送
        Thread {
            for (error in errorsToSend) {
                sendErrorReport(error)
            }
        }.start()
    }

    /**
     * 发送单个错误报告
     */
    private fun sendErrorReport(error: ErrorReport) {
        try {
            val url = URL("${serverUrl}${API_ENDPOINT}")
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Accept", "application/json")
                connectTimeout = 5000
                readTimeout = 5000
                doOutput = true
            }

            val jsonBody = JSONObject().apply {
                put("device_id", deviceId)
                put("type", error.type)
                put("error_message", error.message)
                error.playlistId?.let { put("playlist_id", it) }
                error.mediaId?.let { put("media_id", it) }
                error.additionalData?.let { data ->
                    put("additional_data", JSONObject(data))
                }
            }

            connection.outputStream.use { os ->
                os.write(jsonBody.toString().toByteArray(Charsets.UTF_8))
            }

            val responseCode = connection.responseCode
            if (responseCode in 200..299) {
                Log.d(TAG, "Error report sent successfully: ${error.type}")
            } else {
                Log.e(TAG, "Failed to send error report: HTTP $responseCode")
                // 重新加入队列
                errorQueue.offer(error)
            }

            connection.disconnect()
        } catch (e: Exception) {
            Log.e(TAG, "Exception sending error report: ${e.message}")
            // 重新加入队列
            errorQueue.offer(error)
        }
    }

    /**
     * 启动定时刷新
     */
    private fun startFlushTimer() {
        executor.scheduleWithFixedDelay({
            try {
                flush()
            } catch (e: Exception) {
                Log.e(TAG, "Error in flush timer: ${e.message}")
            }
        }, FLUSH_INTERVAL_MS, FLUSH_INTERVAL_MS, TimeUnit.MILLISECONDS)
    }

    /**
     * 停止服务
     */
    fun stop() {
        flush()
        executor.shutdown()
        try {
            if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
                executor.shutdownNow()
            }
        } catch (e: InterruptedException) {
            executor.shutdownNow()
        }
    }

    /**
     * 获取队列长度
     */
    fun getQueueSize(): Int = errorQueue.size
}

File: frontend/src/player/services/ErrorReporter.ts
/**
 * 错误上报服务
 * 统一收集和上报播放端错误
 */
import axios from 'axios';

// 错误类型定义
export type ErrorType =
    | 'download_failed'
    | 'playback_error'
    | 'cache_error'
    | 'network_error'
    | 'playlist_sync_failed'
    | 'registration_failed'
    | 'insufficient_storage'
    | 'switch_failed'
    | 'media_load_error'
    | 'unknown_error';

// 错误上报数据结构
interface ErrorReport {
    device_id: string;
    type: ErrorType;
    message: string;
    playlist_id?: number | null;
    media_id?: number | null;
    timestamp: string;
    additional_data?: Record<string, unknown>;
}

// 错误上报配置
interface ErrorReporterConfig {
    enabled: boolean;
    maxRetries: number;
    retryDelay: number;
    batchSize: number;
    flushInterval: number;
}

// 默认配置
const DEFAULT_CONFIG: ErrorReporterConfig = {
    enabled: true,
    maxRetries: 3,
    retryDelay: 5000,
    batchSize: 10,
    flushInterval: 30000
};

/**
 * 错误上报服务类
 */
class ErrorReporterService {
    private deviceId: string | null = null;
    private config: ErrorReporterConfig;
    private errorQueue: ErrorReport[] = [];
    private flushTimer: ReturnType<typeof setInterval> | null = null;

    constructor(config: Partial<ErrorReporterConfig> = {}) {
        this.config = { ...DEFAULT_CONFIG, ...config };
    }

    /**
     * 初始化错误上报服务
     */
    init(deviceId: string): void {
        this.deviceId = deviceId;
        this.startFlushTimer();

        // 监听页面关闭，发送剩余错误
        window.addEventListener('beforeunload', () => {
            this.flush(true);
        });
    }

    /**
     * 设置设备 ID
     */
    setDeviceId(deviceId: string): void {
        this.deviceId = deviceId;
    }

    /**
     * 上报错误
     */
    report(
        type: ErrorType,
        message: string,
        options: {
            playlistId?: number | null;
            mediaId?: number | null;
            additionalData?: Record<string, unknown>;
        } = {}
    ): void {
        if (!this.config.enabled || !this.deviceId) {
            console.debug('ErrorReporter disabled or no device ID');
            return;
        }

        const errorReport: ErrorReport = {
            device_id: this.deviceId,
            type,
            message,
            playlist_id: options.playlistId ?? null,
            media_id: options.mediaId ?? null,
            timestamp: new Date().toISOString(),
            additional_data: options.additionalData
        };

        console.warn(`[ErrorReporter] ${type}: ${message}`);

        // 添加到队列
        this.errorQueue.push(errorReport);

        // 如果队列达到批量大小，立即发送
        if (this.errorQueue.length >= this.config.batchSize) {
            this.flush();
        }
    }

    /**
     * 上报下载失败
     */
    reportDownloadFailed(
        message: string,
        playlistId?: number,
        mediaId?: number,
        additionalData?: Record<string, unknown>
    ): void {
        this.report('download_failed', message, {
            playlistId,
            mediaId,
            additionalData
        });
    }

    /**
     * 上报播放错误
     */
    reportPlaybackError(
        message: string,
        mediaId?: number,
        additionalData?: Record<string, unknown>
    ): void {
        this.report('playback_error', message, {
            mediaId,
            additionalData
        });
    }

    /**
     * 上报缓存错误
     */
    reportCacheError(
        message: string,
        additionalData?: Record<string, unknown>
    ): void {
        this.report('cache_error', message, { additionalData });
    }

    /**
     * 上报网络错误
     */
    reportNetworkError(
        message: string,
        additionalData?: Record<string, unknown>
    ): void {
        this.report('network_error', message, { additionalData });
    }

    /**
     * 上报播放列表同步失败
     */
    reportPlaylistSyncFailed(
        message: string,
        playlistId?: number,
        additionalData?: Record<string, unknown>
    ): void {
        this.report('playlist_sync_failed', message, {
            playlistId,
            additionalData
        });
    }

    /**
     * 上报存储空间不足
     */
    reportInsufficientStorage(
        requiredBytes: number,
        availableBytes: number
    ): void {
        this.report('insufficient_storage',
            `Storage insufficient: required ${requiredBytes}, available ${availableBytes}`, {
                additionalData: { requiredBytes, availableBytes }
            });
    }

    /**
     * 刷新错误队列到服务器
     */
    async flush(useBeacon: boolean = false): Promise<void> {
        if (this.errorQueue.length === 0 || !this.deviceId) {
            return;
        }

        const errorsToSend = [...this.errorQueue];
        this.errorQueue = [];

        try {
            if (useBeacon && navigator.sendBeacon) {
                // 使用 sendBeacon 发送（页面关闭时）
                const blob = new Blob([JSON.stringify({
                    device_id: this.deviceId,
                    errors: errorsToSend
                })], { type: 'application/json' });
                navigator.sendBeacon('/api/player/devices/notifications', blob);
            } else {
                // 正常发送
                // 批量发送错误
                for (const error of errorsToSend) {
                    await this.sendWithRetry({
                        device_id: error.device_id,
                        type: error.type as string,
                        playlist_id: error.playlist_id,
                        error_message: error.message
                    });
                }
            }
        } catch (error) {
            console.error('[ErrorReporter] Failed to flush errors:', error);
            // 将失败的错误重新加入队列
            this.errorQueue.unshift(...errorsToSend);
        }
    }

    /**
     * 带重试的发送
     */
    private async sendWithRetry(payload: Record<string, unknown>, retryCount: number = 0): Promise<void> {
        try {
            await axios.post('/api/player/devices/notifications', payload, {
                timeout: 5000
            });
        } catch (error) {
            if (retryCount < this.config.maxRetries) {
                await new Promise(resolve => setTimeout(resolve, this.config.retryDelay));
                await this.sendWithRetry(payload, retryCount + 1);
            } else {
                console.error('[ErrorReporter] Max retries exceeded for error report');
            }
        }
    }

    /**
     * 启动定时刷新
     */
    private startFlushTimer(): void {
        if (this.flushTimer) {
            clearInterval(this.flushTimer);
        }
        this.flushTimer = setInterval(() => {
            this.flush();
        }, this.config.flushInterval);
    }

    /**
     * 停止服务
     */
    stop(): void {
        if (this.flushTimer) {
            clearInterval(this.flushTimer);
            this.flushTimer = null;
        }
        this.flush(true);
    }

    /**
     * 获取当前队列长度
     */
    getQueueLength(): number {
        return this.errorQueue.length;
    }
}

// 导出单例
export const ErrorReporter = new ErrorReporterService();
export default ErrorReporter;
